### DISCUSSION TITLE
Layers and Feature Flags

### DISCUSSION THREAD

**Discussion by ianphil** [Azure/osdu-developer#105] [Ideas]
@danielscholl I see a couple of scenarios a developer might want to try out. 

1. Deploy just the infra (Aks, etc)
2. Deploy infra with middleware (istio, etc)
3. Deploy infra with middleware and service

Do we provide an easy way to do this?

**Comment by danielscholl** [comment]
The capability exists but with different levels of "easy."

**How can I deploy just infrastructure and no software?**

To deploy Infrastructure only set the parameter option called clusterSoftware enable flag to false which will prevent software from loading.  If deployment via "Deploy to Azure" button doing this is easy.  If using AZD it would require you to manually modify the parameter json file.  There currently is no azd Environment feature flag yet to do this automatically.  But it was anticipated to be added.

"clusterSoftware": {
      "value": {
        "enable": true,
        "repository": "${SOFTWARE_REPOSITORY}",
        "branch": "${SOFTWARE_BRANCH}",
        "tag": "${SOFTWARE_TAG}"
      }
    }

**How can I deploy middleware software but not the services?**

This is a little harder to do but can also be done fairly easy.  Software is currently divided into _components_ and _applications_.  Components are the name used in this solution for middleware layers.  ie: Istio, Elastic Certificate Management.  Applications are currently a single item.  All applications or none.   The current design was to perhaps divide OSDU into pieces.  _OSDU-CORE_: The base platform of osdu that is necessary for all parts of OSDU.  (Partition, Legal, Storage, Indexer)   This is the application that is currently underway. Then additional application layers can get added and allow an easy configuration of what application layers are desired to be brought online.

To only install components the following bicep code would have to be commented out from the blade_service.bicep file (line 693-702) prior to running azd provision.

applications: {
        path: serviceLayerConfig.gitops.applications
        dependsOn: [
          'components'
        ]
        timeoutInSeconds: 300
        syncIntervalInSeconds: 300
        retryIntervalInSeconds: 300
        prune: true
      }
